### ISSUE TITLE
Scrolling past end of document using down arrow

### ISSUE THREAD

**Issue by fakhir** [marktext/marktext#1119]
### Description

I'm not sure whether this is intentional or a bug but this behaviour feels very unnatural. If you're editing a document and reach the last line, pressing the down arrow will add a new line to the document and keeping the down arrow pressed will keep adding new lines. Usually the expected behaviour is that if the down arrow is pressed the scrolling will stop on the last line of the document and new lines can only be added using the Enter key.

- [x] Can you reproduce the issue? 

### Steps to reproduce

1. Create any document.
2. Go to its last line and then keep the down arrow key pressed.

**Expected behavior:**

Scrolling should stop at the last line.

**Actual behavior:**

Scrolling continues and new lines keep being added.

### Versions

- Mark Text: 0.14 AppImage and also tried the develop branch in the Git repo (79f912332)
- Operating system: Ubuntu 18.04 (node v10.16.0).

**Comment by fakhir** [comment]
![scroll_bug](https://user-images.githubusercontent.com/1925612/59961129-8a993580-94ec-11e9-82de-f6809cfd475b.gif)

**Comment by LuckyHookin** [comment]
🙋‍♂️
![image](https://user-images.githubusercontent.com/45350455/60165616-2ea71780-9832-11e9-9d34-164c79a1e786.png)

**Comment by LuckyHookin** [comment]
🙋‍♂️

![image](https://user-images.githubusercontent.com/45350455/60165826-99585300-9832-11e9-8fbd-21c122cde428.png)

**Comment by LuckyHookin** [comment]
🙋‍♂️
![image](https://user-images.githubusercontent.com/45350455/60166323-872ae480-9833-11e9-9290-559292234692.gif)

**Comment by Jocs** [comment]
>I'm not sure whether this is intentional or a bug but this behaviour feels very unnatural.

This is the expected behavior, but I think this can be optimized.

> Why need to use Arrow Down or Click bellow last paragraph to create a new paragraph?

You can not create a new paragraph by Enter when the cursor is in a code block.

>Another aspect is that when the scroll occurs, only the scroll bar expands and the cursor goes out of the screen. Only typing something brings the cursor back into view.

Do you mean when you type `Enter` at the last line, the cursor did not scroll into view?

**Comment by fakhir** [comment]
>> Another aspect is that when the scroll occurs, only the scroll bar expands and the cursor goes out of the screen. Only typing something brings the cursor back into view.
>
> Do you mean when you type Enter at the last line, the cursor did not scroll into view?

No, I mean that if i keep the down arrow pressed, the cursor keeps going down and it goes out of view. The scrollbar starts to get smaller indicating that new lines are being added to the end of the document but I cannot see the cursor anymore. If i press Enter then the cursor comes back in view. I have also shown this behaviour in the screen animation which I had posted above. When I press down key the cursor goes out of screen and when I press 'a' it comes back into the screen.

**Comment by adrienbeau** [comment]
I came here to file this issue, only to discover it was already opened.

Creating an endless stream of empty lines at the end of the document is midly infuriating. It can be triggered all too easily (an aging keyboard battery in my case), and there is no easy way to notice these additional lines, since they are not particularly highlighted.

As an author, I do not feel in control of the editor, it feels it can slip from my fingers... Not a major issue but an irritation.

However, I think the down arrow is an excellent shortcut for opening paragraphs in the cases Jocs mentioned, so it should be kept for that usage.

Maybe all that need to be fixed is "if I am on an empty paragraph at the end of the document, then the down arrow does nothing".

**Comment by AamerShareef** [comment]
I was about to raise this as a bug - just saw that this has already in place. Plus one from me on this issue!
Thank you.

**Comment by AamerShareef** [comment]
Just for everyone's visibility on a work around regarding this - you can use page down and page up buttons which can enable scrolling via keyboard - but using arrow keys would be ideal! :)

**Comment by saveriomiroddi** [comment]
I think this a bug, for two reasons:

1. in editor mode, the behavior is different from view mode; there is no scrolling past the last line, when pressing arrow down;
2. in view mode, going past the last line (with arrow down) adds lines, but when switching to editor mode, the lines are not added

I've also never seen this behavior in any other editor.

**Comment by acicovic** [comment]
I also thought it was a bug as I've never have come across any editor that does this, so I think that optimizing this is a good idea.

**Comment by smithtim** [comment]
This also feels like a bug to me, for the reasons mentioned by @saveriomiroddi.

I hope it can be fixed, since it was one of the first things I noticed.

**Comment by Jocs** [comment]
Thank you for the detailed report! This behavior has been fixed in the new editor engine (@muyajs/core). Pressing ArrowDown at the end of the document now creates at most one new empty paragraph (which is intentional, e.g. to exit a code block), but once the last block is already empty, further ArrowDown presses are a no-op — the unlimited paragraph creation is gone. The cursor-out-of-view issue is also fixed: the desktop shell now automatically scrolls the cursor into view on every selection change. Closing as resolved by the engine migration.
